### PR TITLE
Add function name lookup for c++ functions.

### PR DESCRIPTION
**Pull request by pichristoph**
**add function name lookup for c++ functions**
When compiled with Clang++, C++ functions have only been resolved by their linkage name (i.e. when calling a C(++) function from javascript, one had to call them by their linkage names). Now, an additional lookup is done to also look for the C++ function name (such that calling a C++ function from javascript/... by its 'real' name is enabled).

### CODE REVIEW

File: sulong/projects/com.oracle.truffle.llvm.parser/src/com/oracle/truffle/llvm/parser/LazyToTruffleConverterImpl.java [hunk 1]
```diff
@@ -508,4 +508,8 @@ private void copyStructArgumentsToFrame(List<LLVMStatementNode> initializers, No
 
         return formalParamInits;
     }
+
+    public void resolveLinkageName() {
+        parser.parseLinkageName(runtime);
+    }
```
Comment (zapster) [inline sulong/projects/com.oracle.truffle.llvm.parser/src/com/oracle/truffle/llvm/parser/LazyToTruffleConverterImpl.java:514]: I think this is not needed and makes it harder to understand. The caller (`LLVMParser`) can call `parser.parseLinkageName(runtime)` directly.

File: sulong/projects/com.oracle.truffle.llvm.parser/src/com/oracle/truffle/llvm/parser/listeners/FunctionMDOnly.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.oracle.truffle.llvm.parser.listeners;
+
+import com.oracle.truffle.llvm.parser.model.IRScope;
+import com.oracle.truffle.llvm.parser.model.functions.FunctionDefinition;
+import com.oracle.truffle.llvm.parser.scanner.Block;
+import com.oracle.truffle.llvm.parser.scanner.RecordBuffer;
+
+public final class FunctionMDOnly implements ParserListener {
```
Comment (zapster) [inline sulong/projects/com.oracle.truffle.llvm.parser/src/com/oracle/truffle/llvm/parser/listeners/FunctionMDOnly.java:37]: I think we should have Javadoc that mentions that  `MetadataSubprogramOnly` throws.

File: sulong/projects/com.oracle.truffle.llvm.parser/src/com/oracle/truffle/llvm/parser/listeners/FunctionMDOnly.java [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+                return new MetadataSubprogramOnly(types, scope);
+
+            default:
+                return ParserListener.DEFAULT;
+        }
+    }
+
+    @Override
+    public void exit() {
+        // no linkageName found
```
Comment (zapster) [inline sulong/projects/com.oracle.truffle.llvm.parser/src/com/oracle/truffle/llvm/parser/listeners/FunctionMDOnly.java:71]: I think we should mention why it is ok that exit is never called in the case where `MetadataSubprogramOnly` throws.

File: sulong/projects/com.oracle.truffle.llvm.parser/src/com/oracle/truffle/llvm/parser/listeners/MetadataSubprogramOnly.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+ */
+package com.oracle.truffle.llvm.parser.listeners;
+
+import com.oracle.truffle.llvm.parser.metadata.MDBaseNode;
+import com.oracle.truffle.llvm.parser.metadata.MDString;
+import com.oracle.truffle.llvm.parser.metadata.MDSubprogram;
+import com.oracle.truffle.llvm.parser.model.IRScope;
+import com.oracle.truffle.llvm.parser.scanner.RecordBuffer;
+
+public class MetadataSubprogramOnly extends Metadata {
```
Comment (zapster) [inline sulong/projects/com.oracle.truffle.llvm.parser/src/com/oracle/truffle/llvm/parser/listeners/MetadataSubprogramOnly.java:38]: Please add Javadoc that describes what this class is used for (finding `MDSubprogram`) and how the result is returned (via throwing `MDSubprogramParserException`).

File: sulong/projects/com.oracle.truffle.llvm.parser/src/com/oracle/truffle/llvm/parser/model/functions/FunctionDefinition.java [hunk 5]
```diff
@@ -91,6 +93,21 @@ public String getSourceName() {
         return SourceFunction.DEFAULT_SOURCE_NAME.equals(scopeName) ? null : scopeName;
     }
 
+    public String getDisplayName() {
+        // for LLVM from C++, function.name stores the linkage name, but not 'original' C++ name
```
Comment (zapster) [inline sulong/projects/com.oracle.truffle.llvm.parser/src/com/oracle/truffle/llvm/parser/model/functions/FunctionDefinition.java:97]: For readability, I'd change `for LLVM from C++, ...` to `for LLVM code produced from C++ sources`

File: sulong/projects/com.oracle.truffle.llvm.parser/src/com/oracle/truffle/llvm/parser/model/functions/LazyFunctionParser.java [hunk 6]
```diff
@@ -77,4 +79,23 @@ public void parse(DebugInfoFunctionProcessor diProcessor, Source bitcodeSource,
...
+
+    public void parseLinkageName(LLVMParserRuntime runtime) {
+        synchronized (scope) {
+            FunctionMDOnly parser = new FunctionMDOnly(scope, types, function);
+            try {
+                parser.setupScope();
+                scanner.scanBlock(parser);
+            } catch (MDSubprogramParsedException e) {
+                // if linkageName/displayName is found, an exception is thrown (s.t.
+                // parsing/searching does not have to be continued)
```
Comment (zapster) [inline sulong/projects/com.oracle.truffle.llvm.parser/src/com/oracle/truffle/llvm/parser/model/functions/LazyFunctionParser.java:91]: Multiline comments (`/* */`) and proper sentences to make the formatter happy.

File: sulong/tests/com.oracle.truffle.llvm.tests/src/com/oracle/truffle/llvm/tests/interop/MembersTest.java [hunk 7]
```diff
@@ -41,19 +41,26 @@
 public class MembersTest extends InteropTestBase {
```
Comment (zapster) [inline sulong/tests/com.oracle.truffle.llvm.tests/src/com/oracle/truffle/llvm/tests/interop/MembersTest.java:41]: I'd create a new test class for testing C++ members (e.g. `CxxMembersTest`). Not a blocker, though.

File: sulong/projects/com.oracle.truffle.llvm.parser/src/com/oracle/truffle/llvm/parser/LLVMParser.java [hunk 8]
```diff
@@ -126,6 +128,10 @@ private void defineFunction(FunctionSymbol functionSymbol, ModelModule model, Da
         LLVMFunction llvmFunction = LLVMFunction.create(functionSymbol.getName(), library, function, functionSymbol.getType(), runtime.getBitcodeID(), functionSymbol.getIndex(),
                         functionDefinition.isExported());
         runtime.getFileScope().register(llvmFunction);
+        final boolean cxxInterop = LLVMLanguage.getContext().getEnv().getOptions().get(SulongEngineOption.CXX_INTEROP);
```
Comment (rschatz) [inline sulong/projects/com.oracle.truffle.llvm.parser/src/com/oracle/truffle/llvm/parser/LLVMParser.java:131]: You can use `parserRuntime.getContext()` here, that's more efficient. `LLVMLanguage.getContext()` needs to look up the context from a thread-local.

File: sulong/projects/com.oracle.truffle.llvm.parser/src/com/oracle/truffle/llvm/parser/LLVMParser.java [hunk 9]
```diff
@@ -53,6 +53,8 @@
 import java.util.ArrayList;
 import java.util.List;
 
+import org.antlr.v4.runtime.ParserRuleContext;
```
Comment (zapster) [inline sulong/projects/com.oracle.truffle.llvm.parser/src/com/oracle/truffle/llvm/parser/LLVMParser.java:56]: that import looks wrong

**Pull request merged**